include/exadg/incompressible_navier_stokes/spatial_discretization/calculators/wall_shear_stress_calculator.cpp

```diff
@@ -0,0 +1,233 @@
+    *mapping, fe_system, face_quadrature, dealii::update_gradients | dealii::update_normal_vectors);
+#endif
+
+  std::vector<dealii::Tensor<2, dim>> velocity_gradients(fe_face_values.n_quadrature_points);
+  std::vector<dealii::types::global_dof_index> dof_indices(fe_system.dofs_per_cell);
+  const dealii::FEValuesExtractors::Vector     vector(0);
+
+  // fill vector with entries on boundary via cell-loop or prepared container.
+  for(auto const & cell : matrix_free->get_dof_handler(dof_index).active_cell_iterators())
+  {
+    if(cell->is_locally_owned() == true && cell->at_boundary() == true)
+    {
+      cell->get_dof_indices(dof_indices);
+
+#ifdef DEBUG
+      fe_values.reinit(cell);
+      double const tol_sqrd = std::pow((cell->diameter() * rel_tol), 2);
+#endif
+
+      for(auto const face : cell->face_indices())
+      {
+        if(cell->at_boundary(face))
+        {
+          if(write_on_all_boundary_IDs ||
+             write_wall_shear_stress_boundary_IDs.find(cell->face(face)->boundary_id()) !=
+               write_wall_shear_stress_boundary_IDs.end())
+          {
+            fe_face_values.reinit(cell, face);
+            fe_face_values[vector].get_function_gradients(src, velocity_gradients);
+
+            for(auto const i : fe_face_values.quadrature_point_indices())
+            {
+              unsigned int matching_idx = face_to_cell_index[face][i];
+
+#ifdef DEBUG
+              // check point match in debug mode
+              dealii::Point<dim> const point_face = fe_face_values.quadrature_point(i);
+
+              if(point_face.distance_square(fe_values.quadrature_point(matching_idx)) >= tol_sqrd)
+              {
+                matching_idx = dealii::numbers::invalid_dof_index;
+                for(auto const j : fe_values.quadrature_point_indices())
+                {
+                  if(point_face.distance_square(fe_values.quadrature_point(j)) < tol_sqrd)
+                  {
+                    matching_idx = j;
+                    break;
+                  }
+                }
+                AssertThrow(matching_idx == dealii::numbers::invalid_dof_index,
+                            dealii::ExcMessage("Face to cell match recovery not succesful."));
+              }
+#endif
+
+              dealii::Tensor<1, dim> const normal = fe_face_values.normal_vector(i);
+              dealii::Tensor<1, dim> const S_times_n =
+                dynamic_viscosity * (velocity_gradients[i] + transpose(velocity_gradients[i])) *
+                normal;
+              dealii::Tensor<1, dim> const wall_shear_stress =
+                S_times_n - scalar_product((S_times_n), normal) * normal;
+              for(unsigned int d = 0; d < dim; ++d)
+              {
+                if(src.get_partitioner()->in_local_range(
+                     dof_indices[matching_idx + d * dofs_per_component]))
+                  dst(dof_indices[matching_idx + d * dofs_per_component]) = wall_shear_stress[d];
+              }
```

Comment:
This will only work for discontinuous elements such as `FESystem(FE_DGQ)` or `FESystem(FE_SimplexDGP)`, which is something you cannot control here. The reason is that deal.II orders unknowns entity-by-entity, i.e., first all vertices (in their order within a cell), then lines, then faces (aka 'quads'), then volumes (aka 'hexes', this naming also applies to tetrahedra). For discontinuous elements, this ends up being the expected ordering, because all DoFs are associated with the volume, but not for continuous elements where they (necessarily) live on vertices, lines, etc. - so overall you make assumptions on how unknowns are ordered.

Without full knowledge of the code (and all comments, sorry for that), I think that this is a fundamental issue on how this code is structured. As this is post-processing of a quantity that only lives on the boundary, the first question would be whether `DataOutFaces` could have been used (generating a different file). If output in the same file would be desired, the approach I would have taken is to solve a least-squares system that tries to make the polynomial solution representation of a discontinuous solution space (irrespective the underlying `FiniteElement`) **directly in the output module**. This means that I would intuitively not create a global vector first and then write it somewhere, but use the `DataPostprocessor` facilities of `deal.II` for that, see https://www.dealii.org/developer/doxygen/deal.II/classDataPostprocessor.html
Even though there might be some extensions (or performance optimizations) necessary in that module, the overall feeling is that we should not duplicate such functionality in ExaDG unless there are good reasons.